text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,667 @@
+
+expect(stack).to(haveResource('AWS::Foo'));
+```
+
+In such cases, assets will not work because the output directory is only
+determined later (in the `expect` call).
+
+One alternative would be to simply modify these unit tests so that stacks are no
+longer used as roots (basically add `new App()` as the scope). This approach
+would require a change in many unit tests across the code base with no clear
+value to users.
+
+We propose to modify the `Stack` construct such that if a stack is created
+without an explicit `scope`, an `App` instance will automatically be created and
+used as it's scope.
+
+```ts
+const stack = new Stack();
+assert(stack.node.scope instanceof App); // previously it was `undefined`
+```
+
+Since only the root construct may have an empty ID, we will also need to assign
+an ID. We propose to use `"Stack"` since we already have fallback logic that
+uses this as the stack name when the stack does not have an ID (see [TODO]).
+
+This change will allow us to remove any special casing we have for stacks in the
+testing framework and throughout the synthesis code path (we have quite a lot of
+that), because we will be able to assume that `Stage.of(construct.node.root)` is
+never `undefined` and has a `synth()` method which returns a cloud assembly.
+
+Since unit tests sometimes use "incremental tests" for synthesized templates,
+and `stage.synth()` would reuse the synthesized output if called twice, we will
+also need to introduce a `stage.synth({ forceResynth: true })` option. This will
+be the default behavior when using `expect(stack)` or `SynthUtils.synth()`.
+
+The main side effect of this change is that construct paths in unit tests will
+now change. In the above example, `foo.node.path` will change from `MyFoo` to
+`Stack/MyFoo`. Additionally, tests for resources that utilized `node.uniqueId`
+to generate names will also change given `uniqueId` is based on the path.
+
+Since app-less stacks are only used during tests, this should not have
+implications on production code, but it does break some of our test suite.
+
+### Should we do this on 1.x?
+
+In order to reduce [merge conflicts](#repository-migration-efforts) between 1.x
+and 2.x we propose to introduce this change on the 1.x branch prior to forking
+off 2.x.
+
+However, this could be perceived as a breaking change by end-users since
+`node.path` and `node.uniqueId`, and their derivatives, will change for trees
+rooted by a `Stack` and unit tests will need to be updated.
+
+Therefore we propose to introduce this change as a feature flag over the 1.x codebase and enable it in our unit tests (I don't believe we have a way to enable feature flags for all unit tests, but we can devise one).
+
+This will allow us to update our tests in 1.x and avoid the merge conflicts
+forking on 2.x
+
+----
+
+
+
+
+constructs 4.x:
+
+- [x] Migrate to projen
+- [x] Branch to 4.x and release as @next
+- [x] Reintroduce `c.node` instead of `Node.of(c)`
```

Comment:
I kinda liked `Node.of(c)` because it keeps the API of `Construct`/`IConstruct` super clean. However I don't really love how `IConstruct` is just a marker interface and it has no bearing in how `Node.of(IConstruct)` achieves getting the `Node` from there (which makes it unsafe).

Reply:
Exactly, and that's why I am reintroducing `IConstruct.node` so it's not a marker interface anymore. Also, not having `this.node` as a way to access the node is going to break _a lot_ of CDK code.